File: Baekjoon/Level15/9251.cpp
#include<iostream>
#include<vector>
using namespace std;

#define max(x,y) (x > y) ? x : y

int main() {
	string a, b;
	cin >> a >> b;
	vector<vector<int>> t(b.length() + 1, vector<int>(a.length() + 1, 0));
	for (int i = 0; i < b.length(); i++)
	{
		for (int j = 0; j < a.length(); j++) {
			t[i + 1][j + 1] = max(t[i + 1][j], t[i][j + 1]);
			if (a[j] == b[i])
				t[i + 1][j + 1] = max(t[i + 1][j + 1], t[i][j] + 1);
		}
	}
	cout << t[b.length()][a.length()] << '\n';

	return 0;
}

File: Baekjoon/Level15/9251.py
a, b = input(), input()
c, r = len(a), len(b)
t = [[0 for _ in range(c + 1)] for _ in range(r + 1)]
for rr in range(r):
    for cc in range(c):
        if a[cc] == b[rr]:
            t[rr+1][cc+1] = max(t[rr + 1][cc], t[rr][cc + 1], t[rr][cc] + 1)
        else:
            t[rr+1][cc+1] = max(t[rr+1][cc], t[rr][cc+1])

print(t[r][c])
